### PR TITLE
download csv file middle-end

### LINKED ISSUE #193: Download Tables
https://sheetjs.com

- Create a utility function to download data 
- Store this utility in src/utils

### PR DESCRIPTION
**Pull request by SafeDuck**
image

### CODE REVIEW

File: src/components/Admin/Judges.jsx [hunk 1]
```diff
@@ -135,6 +135,8 @@ const Judges = () => {
         objects={filteredJudges}
         filters={filters}
         reset={judges}
+        download={judges}
+        fileName="judges"
```
Comment (shahdivyank) [inline src/components/Admin/Judges.jsx]: Can you add a timestamp to the filenames, this is to prevent having judges-1.csv and such being downloaded. If we were to click download multiple times, ideally it should be timestamped so we know the most recent version. You only need to include day + hour + minute (might remove minute but lets add it for now).

Comment (shahdivyank) [reply src/components/Admin/Judges.jsx]: also it should include the hackathon name, since one person could be in charge of multiple hackathons, we need a way to differentiate between hackathon

File: src/components/Admin/Toolbar.jsx [hunk 2]
```diff
@@ -13,6 +14,8 @@ const Toolbar = ({
   objects,
   filters,
   reset,
+  download,
+  fileName,
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: call the prop: `file`

File: src/components/Admin/Toolbar.jsx [hunk 3]
```diff
@@ -72,7 +75,58 @@ const Toolbar = ({
       );
     }
   };
+  const csvData = objects.map((download) => {
+    const rowData = {};
 
+    if (download.name) {
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: Avoid the if statements, this needs to be more abstract as we may add or delete columns so it needs to work out, for now map through the objects entries and use those values directly. Later we need to add an option to delete/hide particular columns (this is a later issue)

File: src/components/Admin/Toolbar.jsx [hunk 4]
```diff
@@ -110,9 +164,13 @@ const Toolbar = ({
           Reset
         </button>
       </div>
-      <button>
+      <CSVLink
+        data={csvData}
+        filename={`${fileName}.csv`}
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: put the .csv when u initially pass it in at the page level, so like Mentors.jsx and such

**Review by shahdivyank — CHANGES_REQUESTED** [review]

File: src/components/Admin/Toolbar.jsx (hunk 3 above)
Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx]: This will be done by the end of the day

File: src/components/Admin/Toolbar.jsx [hunk 5]
```diff
@@ -62,7 +64,36 @@ const Toolbar = ({
     );
     setToggle(!toggle);
   };
+  // const csvData = objects.entries(download).map(entry => rowData[entry[0]] = entry[1]);
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: remove commented out code

File: src/components/Admin/Judges.jsx [hunk 6]
```diff
@@ -137,6 +137,7 @@ const Judges = () => {
         objects={filteredJudges}
         filters={filters}
         reset={judges}
+        file="JUDGES.csv"
```
Comment (shahdivyank) [inline src/components/Admin/Judges.jsx]: apologies, but now that we r doing the more complex naming scheme, u can keep the .csv at the final stage with the env variables and stuff. just need JUDGES here.

File: src/components/Admin/Toolbar.jsx [hunk 7]
```diff
@@ -62,7 +64,36 @@ const Toolbar = ({
...
+  // const csvData = objects;
+  const csvData = objects.map((download) => {
+    const rowData = {};
+
+    Object.entries(download).forEach(([key, value]) => {
+      if (key === "members" && Array.isArray(value) && value.length > 0) {
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: don't make the assumption that it will be members always, we could have other fields that are arrays and it would work for that as well. u can simply check for the array type and length and then use the join functions

File: src/components/Admin/Toolbar.jsx [hunk 8]
```diff
@@ -62,7 +64,36 @@ const Toolbar = ({
...
+        const memberNames = value.map((member) => member.name).join(", ");
+        const memberEmails = value.map((member) => member.email).join(", ");
+
+        rowData.members = `Names: ${memberNames}, Emails: ${memberEmails}`;
+      } else {
+        if (typeof value === "boolean") {
+          rowData[key] = value ? "Yes" : "No";
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: keep it true and false instead of yes and no

File: src/components/Admin/Toolbar.jsx [hunk 9]
```diff
@@ -62,7 +64,36 @@ const Toolbar = ({
...
+        } else {
+          rowData[key] = value || "";
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: in what situations does this occur?

File: src/components/Admin/Toolbar.jsx [hunk 10]
```diff
@@ -62,7 +64,36 @@ const Toolbar = ({
...
+  // const csvData = objects;
+  const csvData = objects.map((download) => {
+    const rowData = {};
+
+    Object.entries(download).forEach(([key, value]) => {
+      if (key === "members" && Array.isArray(value) && value.length > 0) {
+        const memberNames = value.map((member) => member.name).join(", ");
+        const memberEmails = value.map((member) => member.email).join(", ");
+
+        rowData.members = `Names: ${memberNames}, Emails: ${memberEmails}`;
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: try not to hard code the names and emails, try to use the keys for it

File: src/components/Admin/Toolbar.jsx [hunk 11]
```diff
@@ -62,7 +64,36 @@ const Toolbar = ({
...
+          rowData[key] = value || "";
+        }
+      }
+    });
+
+    return rowData;
+  });
 
+  const currentDate = new Date();
+  const formattedDate = currentDate
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: I have a feeling there is a way to combine the 2 statements into one format

File: src/components/Admin/Toolbar.jsx [hunk 12]
```diff
@@ -103,12 +134,13 @@ const Toolbar = ({
...
         </button>
       </div>
-      <button>
-        <FaDownload
-          size={22.5}
-          className="ml-4 text-hackathon-darkgray hover:opacity-70 duration-150"
-        />
-      </button>
+      <CSVLink
+        data={csvData}
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: call csvData just data

File: src/components/Admin/Toolbar.jsx (hunk 9 above)
Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx]: image
bool is needed

Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx]: otherwise it is blank

Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx]: the last line is needed. Here is what it looks like when you remove both:
image

File: src/components/Admin/Participants.jsx [hunk 13]
```diff
@@ -35,6 +35,8 @@ const Participants = () => {
         setObjects={setParticipants}
         objects={participants}
         filters={filters}
+        reset={participants}
```
Comment (shahdivyank) [inline src/components/Admin/Participants.jsx]: what do you need the reset for?

File: src/components/Admin/Toolbar.jsx [hunk 14]
```diff
@@ -4,9 +4,34 @@ import Checkbox from "../Checkbox";
 import { HiSearch } from "react-icons/hi";
 import Tag from "./Tag.jsx";
 import { FaDownload, FaTrashAlt } from "react-icons/fa";
+import { CSVLink } from "react-csv";
 import { tagColor } from "@/data/Tags";
 
-const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
+function convert(input) {
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: arrow functions

File: src/components/Admin/Toolbar.jsx [hunk 15]
```diff
@@ -71,6 +96,20 @@ const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
     setToggle(!toggle);
   };
 
+  const data = objects.map((items) => {
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: this should be in a function, this is going to run every time react rerenders

File: src/components/Admin/Toolbar.jsx [hunk 16]
```diff
@@ -71,6 +96,20 @@ const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
...
+    const rowData = {};
+    Object.entries(items).map(
+      ([key, value]) => (rowData[key] = convert(value))
+    );
+    return rowData;
+  });
+
+  const formattedDate = new Date()
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: should be in a function

**Comment by SafeDuck** [comment]
Hide the uid, links, dropdown, etc. Needs review now to get pro merged to main

**Review by shahdivyank — CHANGES_REQUESTED** [review]

File: src/components/Admin/Toolbar.jsx [hunk 17]
```diff
@@ -4,9 +4,34 @@ import Checkbox from "../Checkbox";
...
 import { FaDownload, FaTrashAlt } from "react-icons/fa";
+import { CSVLink } from "react-csv";
 import { colors } from "@/data/Tags";
 
-const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
+function convert(input) {
+  if (Array.isArray(input)) {
+    return input.map(convert).join(", ");
+  } else if (typeof input === "boolean") {
+    return input ? "True" : "False";
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: what's the default return? like just return input?

File: src/components/Admin/Toolbar.jsx [hunk 18]
```diff
@@ -4,9 +4,34 @@ import Checkbox from "../Checkbox";
...
+  } else if (typeof input === "object") {
+    return Object.entries(input)
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: can u show an example of this mapping and join, like what does it look like?

File: src/components/Admin/Toolbar.jsx [hunk 19]
```diff
@@ -70,9 +95,25 @@ const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
     );
     setToggle(!toggle);
   };
+  const blacklistArray = ["uid", "selected", "hidden", "links", "dropdown"];
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: just call it blacklist

File: src/components/Admin/Toolbar.jsx [hunk 20]
```diff
@@ -70,9 +95,25 @@ const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
...
+    });
+    return rowData;
+  });
   const handleDelete = () => {
     setObjects(objects.filter((object) => !object.selected));
   };
+  const formattedDate = new Date()
+    .toLocaleDateString("en-US")
+    .replace(/ /g, "_");
+  const formattedTime = new Date()
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: should be a way to combine the 2 formatted date classes

File: src/components/Admin/Toolbar.jsx (hunk 17 above)
Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx]: image

yes, it's on line 22

File: src/components/Admin/Toolbar.jsx (hunk 18 above)
Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx]: it turns {a: 1, b: 2} to "a: 1, b: 2"

Comment (shahdivyank) [reply src/components/Admin/Toolbar.jsx]: - can we just do 1, 2, 3...
- does the generated CSV have a table header?

File: src/components/Admin/Judges.jsx [hunk 21]
```diff
@@ -34,6 +34,8 @@ const Judges = () => {
         setObjects={setJudges}
         objects={judges}
         filters={filters}
+        reset={judges}
```
Comment (shahdivyank) [inline src/components/Admin/Judges.jsx]: remove reset

File: src/components/Admin/Toolbar.jsx (hunk 17 above)
Comment (shahdivyank) [reply src/components/Admin/Toolbar.jsx]: what happens when u just write `return input` in this situation?

File: src/components/Admin/Toolbar.jsx [hunk 22]
```diff
@@ -4,9 +4,34 @@ import Checkbox from "../Checkbox";
...
+  if (Array.isArray(input)) {
+    return input.map(convert).join(", ");
+  } else if (typeof input === "boolean") {
+    return input ? "True" : "False";
+  } else if (typeof input === "object") {
+    return Object.entries(input)
+      .map(([key, value]) => `${key}: ${convert(value)}`)
+      .join(", ");
+  } else if (typeof input === "number") {
+    return input.toString();
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: do we need to convert to a string? does it not work when we have a number?

File: src/components/Admin/Toolbar.jsx (hunk 17 above)
Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx]: image
image

File: src/components/Admin/Toolbar.jsx (hunk 18 above)
Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx]: - I don't know about 1,2,3...
- There is no table header prop that you can pass that I know of. We would need to do some hard coding of the "data" prop, so then we get a "header" in the CSV file

File: src/components/Admin/Toolbar.jsx (hunk 22 above)
Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx]: It was just in case we needed it. According to my comment above with the screenshots, it seems like we don't need it

File: src/components/Admin/Toolbar.jsx (hunk 17 above)
Comment (shahdivyank) [reply src/components/Admin/Toolbar.jsx]: I mean just for boolean types, if you just return input on boolean types, what is the value?

File: src/components/Admin/Toolbar.jsx (hunk 18 above)
Comment (shahdivyank) [reply src/components/Admin/Toolbar.jsx]: headers is getting passed into the table component and the toolbar component, so we should be able to add the headers as the first row

File: src/components/Admin/Toolbar.jsx (hunk 22 above)
Comment (shahdivyank) [reply src/components/Admin/Toolbar.jsx]: remove the toString then, if we dont need it

File: src/components/Admin/Toolbar.jsx (hunk 17 above)
Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx]: tested it, we don't need it

Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx]: just return input

File: src/components/Admin/Toolbar.jsx [hunk 23]
```diff
@@ -4,10 +4,30 @@ import Checkbox from "../Checkbox";
...
 import Tag from "./Tag.jsx";
 import { FaDownload, FaTrashAlt } from "react-icons/fa";
+import { CSVLink } from "react-csv";
 import { colors } from "@/data/Tags";
 import Popup from "./Popup";
 
-const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
+const convert = (input) => {
+  if (Array.isArray(input)) {
+    return input.map(convert).join(", ");
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: dont need the map

File: src/components/Admin/Toolbar.jsx [hunk 24]
```diff
@@ -4,10 +4,30 @@ import Checkbox from "../Checkbox";
...
+  } else if (typeof input === "object") {
+    return Object.values(input).map(convert).join(", ");
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: dont need the map

File: src/components/Admin/Toolbar.jsx [hunk 25]
```diff
@@ -4,10 +4,30 @@ import Checkbox from "../Checkbox";
...
+  } else {
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: remove the else and just have the return

File: src/components/Admin/Toolbar.jsx [hunk 26]
```diff
@@ -72,9 +92,35 @@ const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
     );
     setToggle(!toggle);
   };
+  const blacklist = ["uid", "selected", "hidden", "links", "dropdown"];
+  const mapObjectsToCSVData = (objects, blacklist, headers) => {
+    const data = [];
+    const columnNames = headers
+      .map((header) => header.text)
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: u should be able to combine the map and filter and do it in one go

File: src/components/Admin/Toolbar.jsx [hunk 27]
```diff
@@ -72,9 +92,35 @@ const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
...
+      .filter((key) => key && !blacklist.includes(key));
+    data.push(columnNames);
+    objects.forEach((item) => {
+      const rowData = {};
+      for (const key of columnNames) {
+        const value = item[key];
+        rowData[key] = convert(value);
+      }
+      data.push(Object.values(rowData));
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: if u just need an array, then just store it as an array originally, it should remove the need for Object.values

File: src/components/Admin/Toolbar.jsx [hunk 28]
```diff
@@ -72,9 +92,35 @@ const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
...
+    return data;
+  };
+
+  const data = mapObjectsToCSVData(objects, blacklist, headers);
   const handleDelete = () => {
     setObjects(objects.filter((object) => !object.selected));
   };
+  const date = new Date();
+  const formattedDate = date.toLocaleDateString("en-US").replace(/ /g, "_");
+  const formattedTime = date
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: see if this date format time can be converted with date format time

File: src/components/Admin/Toolbar.jsx [hunk 29]
```diff
@@ -4,10 +4,28 @@ import Checkbox from "../Checkbox";
...
-const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
+const convert = (input) => {
+  if (Array.isArray(input)) {
+    return input.join(", ");
+  } else if (typeof input === "object") {
+    return Object.values(input).join(", ");
+  }
+  return input;
+};
+const Toolbar = ({
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx:20]:
```suggestion

const Toolbar = ({
```

File: src/components/Admin/Toolbar.jsx [hunk 30]
```diff
@@ -72,9 +91,41 @@ const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
...
+    const columns = headers
+      .filter((header) => header.text && !blacklist.includes(header.text))
+      .map((header) => header.text);
+    // divyank's code
+    // const columns = headers.filter((header) => {
+    //   if (header.text && !blacklist.includes(header.text)) {
+    //     return header.text;
+    //   }
+    // });
+    const data = [columns];
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx:106]: why is this needed?

File: src/components/Admin/Toolbar.jsx [hunk 31]
```diff
@@ -72,9 +91,41 @@ const Toolbar = ({ input, setInput, tags, setObjects, objects, filters }) => {
...
   };
+  const date = new Date();
+  const formattedDateTime = date
+    .toLocaleString("en-US", {
+      month: "2-digit",
+      day: "2-digit",
+      year: "numeric",
+      hour: "2-digit",
+      minute: "2-digit",
+      hour12: true,
```
Comment (shahdivyank) [inline src/components/Admin/Toolbar.jsx]: keep it 24 hour, also make sure its UTC time

File: src/components/Admin/Toolbar.jsx (hunk 30 above)
Comment (SafeDuck) [reply src/components/Admin/Toolbar.jsx:106]: It's the same thing as data.push(columns)
It's adding the headers to the top of the csv file

**Review by shahdivyank — APPROVED** [review]

**Review by menthy-wu — APPROVED** [review]
